Address safer cpp warnings in platform/graphics/coretext

cdumez opened the pull request:
#### e980cbd147631cfd2a7e33c395c4cc84f4537d2d
<pre>
Address safer cpp warnings in platform/graphics/coretext
<a href="https://bugs.webkit.org/show_bug.cgi?id=299070">https://bugs.webkit.org/show_bug.cgi?id=299070</a>

Reviewed by Darin Adler.

* Source/WebCore/SaferCPPExpectations/NoUncountedMemberCheckerExpectations:
* Source/WebCore/SaferCPPExpectations/RetainPtrCtorAdoptCheckerExpectations:
* Source/WebCore/SaferCPPExpectations/UncountedLocalVarsCheckerExpectations:
* Source/WebCore/SaferCPPExpectations/UnretainedCallArgsCheckerExpectations:
* Source/WebCore/SaferCPPExpectations/UnretainedLocalVarsCheckerExpectations:
* Source/WebCore/accessibility/cocoa/AccessibilityObjectCocoa.mm:
(WebCore::fontFrom):
* Source/WebCore/accessibility/ios/AccessibilityObjectIOS.mm:
(WebCore::attributeStringSetStyle):
* Source/WebCore/editing/cocoa/AttributedString.mm:
(WebCore::toNSObject):
* Source/WebCore/editing/cocoa/EditingHTMLConverter.mm:
(WebCore::updateAttributes):
* Source/WebCore/editing/cocoa/FontAttributeChangesCocoa.mm:
(WebCore::FontChanges::platformFontFamilyNameForCSS const):
* Source/WebCore/editing/cocoa/FontAttributesCocoa.mm:
(WebCore::FontAttributes::createDictionary const):
* Source/WebCore/editing/cocoa/NodeHTMLConverter.mm:
(_font):
* Source/WebCore/platform/graphics/Font.h:
(WebCore::Font::ctFont const):
(WebCore::Font::protectedCTFont const):
(WebCore::Font::getCTFont const): Deleted.
* Source/WebCore/platform/graphics/FontPlatformData.h:
(WebCore::FontPlatformData::protectedCTFont const):
* Source/WebCore/platform/graphics/GraphicsContext.h:
(WebCore::GraphicsContext::protectedPlatformContext const):
* Source/WebCore/platform/graphics/cocoa/GraphicsContextCocoa.mm:
(WebCore::GraphicsContext::drawMultiRepresentationHEIC):
* Source/WebCore/platform/graphics/coretext/ComplexTextControllerCoreText.mm:
(WebCore::buildCoreTextTypesetterEmbeddingLevelDictionary):
(WebCore::typesetterOptionsSingleton):
(WebCore::ComplexTextController::collectComplexTextRunsForCharacters):
(WebCore::typesetterOptions): Deleted.
* Source/WebCore/platform/graphics/coretext/DrawGlyphsRecorder.cpp:
(WebCore::DrawGlyphsRecorder::recordInitialColors):
(WebCore::DrawGlyphsRecorder::prepareInternalContext):
(WebCore::DrawGlyphsRecorder::updateShadow):
(WebCore::DrawGlyphsRecorder::recordDrawGlyphs):
(WebCore::DrawGlyphsRecorder::recordDrawPath):
(WebCore::DrawGlyphsRecorder::drawNativeText):
* Source/WebCore/platform/graphics/coretext/DrawGlyphsRecorder.h:
* Source/WebCore/platform/graphics/coretext/FontCascadeCoreText.cpp:
(WebCore::FontCascade::primaryFontIsSystemFont const):
* Source/WebCore/platform/graphics/coretext/FontCoreText.cpp:
(WebCore::Font::platformInit):
(WebCore::Font::platformCharWidthInit):
(WebCore::Font::supportsOpenTypeAlternateHalfWidths const):
(WebCore::Font::supportsSmallCaps const):
(WebCore::Font::supportsAllSmallCaps const):
(WebCore::Font::supportsPetiteCaps const):
(WebCore::Font::supportsAllPetiteCaps const):
(WebCore::openTypeFeature):
(WebCore::trueTypeFeature):
(WebCore::defaultSelectorForTrueTypeFeature):
(WebCore::removedFeature):
(WebCore::createCTFontWithoutSynthesizableFeatures):
(WebCore::Font::createFontWithoutSynthesizableFeatures const):
(WebCore::Font::platformCreateScaledFont const):
(WebCore::supportsOpenTypeFeature):
(WebCore::Font::platformCreateHalfWidthFont const):
(WebCore::Font::platformWidthForGlyph const):
(WebCore::Font::applyTransforms const):
(WebCore::Font::determinePitch):
(WebCore::Font::platformBoundsForGlyph const):
(WebCore::Font::platformBoundsForGlyphs const):
(WebCore::Font::platformPathForGlyph const):
(WebCore::Font::platformSupportsCodePoint const):
(WebCore::Font::isProbablyOnlyUsedToRenderIcons const):
(WebCore::Font::otSVGTable const):
(WebCore::Font::hasComplexColorFormatTables const):
(WebCore::Font::glyphsWithComplexColorFormat const):
(WebCore::Font::glyphHasComplexColorFormat const):
(WebCore::Font::metricsForMultiRepresentationHEIC const):
* Source/WebCore/platform/graphics/coretext/FontPlatformDataCoreText.cpp:
(WebCore::FontPlatformData::FontPlatformData):
(WebCore::findFontDescriptor):
(WebCore::createCTFont):
(WebCore::FontPlatformData::create):
(WebCore::FontPlatformData::registeredFont const):
(WebCore::FontPlatformData::objectForEqualityCheck const):
(WebCore::FontPlatformData::openTypeTable const):
(WebCore::FontPlatformData::familyName const):
(WebCore::FontPlatformData::attributes const):
(WebCore::FontPlatformData::fromIPCData):
(WebCore::FontPlatformData::toIPCData const):
(WebCore::FontPlatformSerializedAttributes::fromCF):
(WebCore::FontPlatformOpticalSize::fromCF):
* Source/WebCore/platform/graphics/coretext/GlyphPageCoreText.cpp:
(WebCore::GlyphPage::fill):
* Source/WebKit/UIProcess/mac/WebViewImpl.mm:
(WebKit::WebViewImpl::updateFontManagerIfNeeded):
* Source/WebKit/WebProcess/WebCoreSupport/mac/WebPopupMenuMac.mm:
(WebKit::WebPopupMenu::setUpPlatformData):
* Source/WebKit/WebProcess/WebPage/ios/WebPageIOS.mm:
(WebKit::WebPage::requestAutocorrectionData):
* Source/WebKitLegacy/mac/DOM/DOM.mm:
(-[DOMElement _font]):
* Source/WebKitLegacy/mac/WebCoreSupport/PopupMenuMac.mm:
(PopupMenuMac::populate):
(PopupMenuMac::show):
* Source/WebKitLegacy/mac/WebView/WebFrame.mm:
(-[WebFrame fontForSelection:]):
* Source/WebKitLegacy/mac/WebView/WebHTMLView.mm:
(-[WebHTMLView _updateFontPanel]):

Canonical link: <a href="https://commits.webkit.org/300214@main">https://commits.webkit.org/300214@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/aaa24f152c6f0b99cde50fac0be801703cc0abec

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/121687 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/41388 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/32052 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/128208 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/73777 "Built successfully") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/1c841cad-3093-4e9d-ad97-a02bebcd132f) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/123563 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/42100 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/49979 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/92464 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/61477 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/32c1be8c-ba26-4c6d-9bca-5e9abac9e4fd) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/124639 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/33588 "Passed tests") | [❌ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/108992 "Found 1 new API test failure: TestIPC.ConnectionTest/ConnectionRunLoopTest.SendAsyncAndInvalidate/ClientIsA (failure)") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/73125 "Passed tests") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/4d39c011-d410-4001-acf2-d4088357cfc5) 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/32598 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/27153 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/71727 "Built successfully") | | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/103086 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/27339 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/131002 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/48622 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/36982 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/101033 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/48994 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/105208 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/100924 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/25596 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/46299 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/24411 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/45322 "Built successfully") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/48480 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/54207 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/47949 "Built successfully") | | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/51298 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/49632 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

darinadler commented on Source/WebCore/editing/cocoa/AttributedString.mm:
```diff
@@ -383,7 +383,7 @@ static MultiRepresentationHEICAttachmentData toMultiRepresentationHEICAttachment
         return value;
     }, [] (const Ref<Font>& font) -> RetainPtr<id> {
-        return (__bridge PlatformFont *)(font->getCTFont());
+        return (__bridge PlatformFont *)(font->ctFont());
```
I don't think the extra parentheses that were here are helpful and we can remove them:
```suggestion
        return (__bridge PlatformFont *)font->ctFont();
```

darinadler commented on Source/WebCore/platform/graphics/coretext/ComplexTextControllerCoreText.mm:
```diff
@@ -268,24 +270,24 @@ static CFDictionaryRef typesetterOptions()
                 }
                 if (!runFont) {
-                    RetainPtr<CFStringRef> fontName = adoptCF(CTFontCopyPostScriptName(runCTFont));
+                    RetainPtr<CFStringRef> fontName = adoptCF(CTFontCopyPostScriptName(runCTFont.get()));
```
```suggestion
                    RetainPtr fontName = adoptCF(CTFontCopyPostScriptName(runCTFont.get()));
```

darinadler commented on Source/WebCore/platform/graphics/GraphicsContext.h:
```diff
@@ -86,6 +86,12 @@ class GraphicsContext {
 
     virtual bool hasPlatformContext() const { return false; }
     virtual PlatformGraphicsContext* platformContext() const { return nullptr; }
+#if USE(CG)
+    RetainPtr<PlatformGraphicsContext> protectedPlatformContext() const { return platformContext(); }
+#else
+    // On other platforms, the PlatformGraphicsContext type is not refcounted.
+    PlatformGraphicsContext* protectedPlatformContext() const { return platformContext(); }
+#endif
```
Now that this is platform-specific I'm not sure it needs an abstract name. Might be easier to understand with a concrete one:
```suggestion
#if USE(CG)
    RetainPtr<CGContextRef> protectedCGContext() const { return platformContext(); }
#endif
```

darinadler commented on Source/WebCore/platform/graphics/coretext/DrawGlyphsRecorder.cpp:
```diff
@@ -506,8 +506,9 @@ void DrawGlyphsRecorder::drawNativeText(CTFontRef font, CGFloat fontSize, CTLine
-    CTLineDraw(line, m_internalContext->platformContext());
+    RetainPtr platformContext = m_internalContext->platformContext();
+    CGContextSetTextPosition(platformContext.get(), 0, 0);
+    CTLineDraw(line, platformContext.get());
```
```suggestion
    RetainPtr context = m_internalContext->platformContext();
    CGContextSetTextPosition(context.get(), 0, 0);
    CTLineDraw(line, context.get());
```

darinadler commented on Source/WebCore/platform/graphics/coretext/FontPlatformDataCoreText.cpp:
```diff
@@ -361,15 +361,15 @@ std::optional<FontPlatformSerializedAttributes> FontPlatformSerializedAttributes
 
-    auto paletteColors = static_cast<CFDictionaryRef>(CFDictionaryGetValue(dictionary, kCTFontPaletteColorsAttribute));
-    result.paletteColors = pairExtractor.template operator()<CGColorRef>(paletteColors, CGColorGetTypeID());
+    RetainPtr paletteColors = static_cast<CFDictionaryRef>(CFDictionaryGetValue(dictionary, kCTFontPaletteColorsAttribute));
```
These should use a checked CF cast of some sort rather than `static_cast`. They have handwritten type checking code but that's what checked and dynamic CF casts are now for.

cdumez committed:
Address safer cpp warnings in platform/graphics/coretext
https://bugs.webkit.org/show_bug.cgi?id=299070

Reviewed by Darin Adler.

* Source/WebCore/SaferCPPExpectations/NoUncountedMemberCheckerExpectations:
* Source/WebCore/SaferCPPExpectations/RetainPtrCtorAdoptCheckerExpectations:
* Source/WebCore/SaferCPPExpectations/UncountedLocalVarsCheckerExpectations:
* Source/WebCore/SaferCPPExpectations/UnretainedCallArgsCheckerExpectations:
* Source/WebCore/SaferCPPExpectations/UnretainedLocalVarsCheckerExpectations:
* Source/WebCore/accessibility/cocoa/AccessibilityObjectCocoa.mm:
(WebCore::fontFrom):
* Source/WebCore/accessibility/ios/AccessibilityObjectIOS.mm:
(WebCore::attributeStringSetStyle):
* Source/WebCore/editing/cocoa/AttributedString.mm:
(WebCore::toNSObject):
* Source/WebCore/editing/cocoa/EditingHTMLConverter.mm:
(WebCore::updateAttributes):
* Source/WebCore/editing/cocoa/FontAttributeChangesCocoa.mm:
(WebCore::FontChanges::platformFontFamilyNameForCSS const):
* Source/WebCore/editing/cocoa/FontAttributesCocoa.mm:
(WebCore::FontAttributes::createDictionary const):
* Source/WebCore/editing/cocoa/NodeHTMLConverter.mm:
(_font):
* Source/WebCore/platform/graphics/Font.h:
(WebCore::Font::ctFont const):
(WebCore::Font::protectedCTFont const):
(WebCore::Font::getCTFont const): Deleted.
* Source/WebCore/platform/graphics/FontPlatformData.h:
(WebCore::FontPlatformData::protectedCTFont const):
* Source/WebCore/platform/graphics/GraphicsContext.h:
(WebCore::GraphicsContext::protectedPlatformContext const):
* Source/WebCore/platform/graphics/cocoa/GraphicsContextCocoa.mm:
(WebCore::GraphicsContext::drawMultiRepresentationHEIC):
* Source/WebCore/platform/graphics/coretext/ComplexTextControllerCoreText.mm:
(WebCore::buildCoreTextTypesetterEmbeddingLevelDictionary):
(WebCore::typesetterOptionsSingleton):
(WebCore::ComplexTextController::collectComplexTextRunsForCharacters):
(WebCore::typesetterOptions): Deleted.
* Source/WebCore/platform/graphics/coretext/DrawGlyphsRecorder.cpp:
(WebCore::DrawGlyphsRecorder::recordInitialColors):
(WebCore::DrawGlyphsRecorder::prepareInternalContext):
(WebCore::DrawGlyphsRecorder::updateShadow):
(WebCore::DrawGlyphsRecorder::recordDrawGlyphs):
(WebCore::DrawGlyphsRecorder::recordDrawPath):
(WebCore::DrawGlyphsRecorder::drawNativeText):
* Source/WebCore/platform/graphics/coretext/DrawGlyphsRecorder.h:
* Source/WebCore/platform/graphics/coretext/FontCascadeCoreText.cpp:
(WebCore::FontCascade::primaryFontIsSystemFont const):
* Source/WebCore/platform/graphics/coretext/FontCoreText.cpp:
(WebCore::Font::platformInit):
(WebCore::Font::platformCharWidthInit):
(WebCore::Font::supportsOpenTypeAlternateHalfWidths const):
(WebCore::Font::supportsSmallCaps const):
(WebCore::Font::supportsAllSmallCaps const):
(WebCore::Font::supportsPetiteCaps const):
(WebCore::Font::supportsAllPetiteCaps const):
(WebCore::openTypeFeature):
(WebCore::trueTypeFeature):
(WebCore::defaultSelectorForTrueTypeFeature):
(WebCore::removedFeature):
(WebCore::createCTFontWithoutSynthesizableFeatures):
(WebCore::Font::createFontWithoutSynthesizableFeatures const):
(WebCore::Font::platformCreateScaledFont const):
(WebCore::supportsOpenTypeFeature):
(WebCore::Font::platformCreateHalfWidthFont const):
(WebCore::Font::platformWidthForGlyph const):
(WebCore::Font::applyTransforms const):
(WebCore::Font::determinePitch):
(WebCore::Font::platformBoundsForGlyph const):
(WebCore::Font::platformBoundsForGlyphs const):
(WebCore::Font::platformPathForGlyph const):
(WebCore::Font::platformSupportsCodePoint const):
(WebCore::Font::isProbablyOnlyUsedToRenderIcons const):
(WebCore::Font::otSVGTable const):
(WebCore::Font::hasComplexColorFormatTables const):
(WebCore::Font::glyphsWithComplexColorFormat const):
(WebCore::Font::glyphHasComplexColorFormat const):
(WebCore::Font::metricsForMultiRepresentationHEIC const):
* Source/WebCore/platform/graphics/coretext/FontPlatformDataCoreText.cpp:
(WebCore::FontPlatformData::FontPlatformData):
(WebCore::findFontDescriptor):
(WebCore::createCTFont):
(WebCore::FontPlatformData::create):
(WebCore::FontPlatformData::registeredFont const):
(WebCore::FontPlatformData::objectForEqualityCheck const):
(WebCore::FontPlatformData::openTypeTable const):
(WebCore::FontPlatformData::familyName const):
(WebCore::FontPlatformData::attributes const):
(WebCore::FontPlatformData::fromIPCData):
(WebCore::FontPlatformData::toIPCData const):
(WebCore::FontPlatformSerializedAttributes::fromCF):
(WebCore::FontPlatformOpticalSize::fromCF):
* Source/WebCore/platform/graphics/coretext/GlyphPageCoreText.cpp:
(WebCore::GlyphPage::fill):
* Source/WebKit/UIProcess/mac/WebViewImpl.mm:
(WebKit::WebViewImpl::updateFontManagerIfNeeded):
* Source/WebKit/WebProcess/WebCoreSupport/mac/WebPopupMenuMac.mm:
(WebKit::WebPopupMenu::setUpPlatformData):
* Source/WebKit/WebProcess/WebPage/ios/WebPageIOS.mm:
(WebKit::WebPage::requestAutocorrectionData):
* Source/WebKitLegacy/mac/DOM/DOM.mm:
(-[DOMElement _font]):
* Source/WebKitLegacy/mac/WebCoreSupport/PopupMenuMac.mm:
(PopupMenuMac::populate):
(PopupMenuMac::show):
* Source/WebKitLegacy/mac/WebView/WebFrame.mm:
(-[WebFrame fontForSelection:]):
* Source/WebKitLegacy/mac/WebView/WebHTMLView.mm:
(-[WebHTMLView _updateFontPanel]):

Canonical link: https://commits.webkit.org/300214@main
